Enable malware scanner - Defender for Storage

anatbal opened the pull request:
# Resolves #3768

## What is being addressed

Since Defender for storage has become publicly available, it required many changes to the existing airlock **import** infra in order to make it work. This PR automates the process, but here is a complete description of the process in order for customers that are already using it can enable it themselves.

1.  First, in the config.yaml, `enable_airlock_malware_scanning: true` must be set.
2. The Microsoft Defender for Cloud should be enabled on "in progress" storage account, stalimip<tre-id>.
3. When enabling it, set the settings as the following: 
![image](https://github.com/microsoft/AzureTRE/assets/13421451/70213f0b-a9b5-496b-8d19-e6fd0194d9e2)
it's important to notice the event grid custom topic is publicly accessed since currently the scan results cannot be delivered over private endpoint.
4. If you decide to create the custom event grid manually, there must be a subscription for a service bus queue that is called "scan-result". This will trigger the azure function to act upon the scanner results. 

- This solution was tested on a clean deployment and verified for both import and export
@migldasilva, let me know if it's clear

anatbal commented:
/test-extended

migldasilva commented:
@anatbal I have just reviewed the code and observed that a private endpoint is not being created for Scanresult eventgrid topic. Is there a specific reason for that?

tamirkamara reviewed (approved):
Thanks for adding this. Just a few comments I think we need to sort.

anatbal commented on airlock_processor/ScanResultTrigger/__init__.py:
```diff
@@ -14,7 +14,7 @@ def main(msg: func.ServiceBusMessage,
     logging.info("Python ServiceBus queue trigger processed message - Malware scan result arrived!")
     body = msg.get_body().decode('utf-8')
-    logging.info('Python ServiceBus queue trigger processed message: %s', body)
+    logging.info('Python ServiceBus queue trigger processed message: {0}'.format(body))
```
fixed

anatbal commented on airlock_processor/ScanResultTrigger/__init__.py:
```diff
@@ -34,9 +34,9 @@ def main(msg: func.ServiceBusMessage,
+        verdict = json_body["data"]["scanResultType"]
     except KeyError as e:
-        logging.error("body was not as expected {}", e)
+        logging.error(f'body was not as expected {e}')
```
returned it to be as it was

anatbal commented on core/terraform/variables.tf:
```diff
@@ -130,6 +130,13 @@ variable "arm_environment" {
+  description = "Used as an environment variable to determine if Terraform should use a managed identity"
+}
+
+
```
Changed it and removed the comment

anatbal commented on core/terraform/main.tf:
```diff
@@ -17,11 +17,19 @@ terraform {
     }
+    azapi = {
+      source  = "Azure/azapi"
+      version = "=1.9.0"
```
fixed

anatbal commented on core/version.txt:
```diff
@@ -1 +1 @@
-__version__ = "0.8.9"
+__version__ = "0.8.10"
```
fixed

anatbal commented on core/version.txt:
```diff
@@ -1 +1 @@
-__version__ = "0.8.9"
+__version__ = "0.8.10"
```
fixed

anatbal commented on airlock_processor/ScanResultTrigger/__init__.py:
```diff
@@ -46,10 +46,10 @@ def main(msg: func.ServiceBusMessage,
     completed_step = constants.STAGE_SUBMITTED
     if verdict == constants.NO_THREATS:
-        logging.info('No malware were found in request id %s, moving to %s stage', request_id, constants.STAGE_IN_REVIEW)
+        logging.info('No malware were found in request id {0}, moving to {1} stage'.format(request_id, constants.STAGE_IN_REVIEW))
```
fixed

anatbal commented:
> @anatbal I have just reviewed the code and observed that a private endpoint is not being created for Scanresult eventgrid topic. Is there a specific reason for that?

Hey @migldasilva. Yes, since the scan result publishing is not supported over private endpoints

anatbal commented:
/test-extended

migldasilva commented:
@anatbal I have just deployed the this code and observed a small change in the overall behavior. After testing the whole import cycle, the blob container created inside `stalimex<MY_ENVIRONMENT>` is not deleted. Is not such containers expected to be deleted after moving files?

anatbal commented:
> @anatbal I have just deployed the this code and observed a small change in the overall behavior. After testing the whole import cycle, the blob container created inside `stalimex<MY_ENVIRONMENT>` is not deleted. Is not such containers expected to be deleted after moving files?

You are right. I will dig into it and provide a code fix for that as well.

tamirkamara commented on core/terraform/main.tf:
```diff
@@ -17,11 +17,19 @@ terraform {
 }
 
+provider "azapi" {
+  use_msi = var.arm_use_msi
```
Well, I don't know about that but at any case...
If there's an environment variable ARM_USE_MSI then the providers will use it without the need to explicitly include them in the definition. So I think we can remove this line.

marrobi replied:
I think the issue is it defaults to True, so needs to be explicitly set somewhere. Not being defined/optional is an issue.

marrobi reviewed (approved):
LGTM in principal, will try find time to test it out.

Can you update the docs to describe the setting and malware scanning setting etc where appropriate? Thanks.

anatbal commented on core/terraform/airlock/storage_accounts.tf:
```diff
@@ -111,6 +111,32 @@ resource "azurerm_storage_account" "sa_import_in_progress" {
+        scanResultsEventGridTopicResourceId = azurerm_eventgrid_topic.scan_result[0].id
+      }
+      sensitiveDataDiscovery = {
+        isEnabled = true
```
disabled

anatbal commented:
@migldasilva I have fixed the deletion problem, the final version of this PR is also deleting the container as it should, and was tested.
@marrobi  I added additional documenntation where it was needed

anatbal commented:
/test-extended

anatbal commented:
/tre-destroy-env

anatbal commented:
/test-destroy-env

anatbal commented:
/test-extended

migldasilva commented on airlock_processor/BlobCreatedTrigger/__init__.py:
```diff
@@ -32,10 +32,11 @@ def main(msg: func.ServiceBusMessage,
             # If malware scanning is enabled, the fact that the blob was created can be dismissed.
             # It will be consumed by the malware scanning service
             logging.info('Malware scanning is enabled. no action to perform.')
+            send_delete_event(dataDeletionEvent, json_body, request_id)
```
I have tested this version and it's triggering a runtime error. The blob storage inside `stalimip<TRE_ID>` is being deleted twice. I could stop getting such errors moving this line to a `if` statement. Something like this:

```
if event_type is None or event_type != 'Microsoft.Storage.BlobDeleted':
    send_delete_event(dataDeletionEvent, json_body, request_id)
return
```

A few lines above I checked if `eventType` is in the `json_body`, and if so, I assigned it the value store in the `json_body`.

I believe the deletion occur because, 1) the file is moved from storage account `stalimip<TRE_ID>` to a storage account inside the workspace, and it's the last blob in the storage account, and 2) the request status changed (I couldn't identify the correct order of execution).

anatbal replied:
@migldasilva  I will test this in a separate PR. Will try to reproduce on my end, since this shouldnt be triggered from a 'Blob.Deleted' event....

marrobi commented:
/test-destroy-env
